File: design-pattern/composite-pattern/composite.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class FileSystem
{
public:
    virtual void ls(int indent = 0) = 0;
    virtual void openAll(int indent = 0) = 0;
    virtual string getName() = 0;
    virtual int getSize() = 0;
    virtual bool isFolder() = 0;
    virtual FileSystem *changeFolder(const string &name) = 0;
    virtual ~FileSystem() {}
};

class File : public FileSystem
{
private:
    string name;
    int size;

public:
    File(string name, int size) : name(name), size(size) {}

    void ls(int indent) override
    {
        cout << string(indent, ' ') << "File: " << name
             << " (" << size << " bytes)" << endl;
    }

    void openAll(int indent) override
    {
        cout << string(indent, ' ') << "Opening file: " << name << endl;
    }

    string getName() override
    {
        return name;
    }

    int getSize() override
    {
        return size;
    }

    bool isFolder() override
    {
        return false;
    }

    FileSystem *changeFolder(const string &) override
    {
        return nullptr;
    }
};

class Folder : public FileSystem
{
private:
    string name;
    vector<FileSystem *> fs;

public:
    Folder(string name) : name(name) {}

    void addItem(FileSystem *f)
    {
        fs.push_back(f);
    }

    void ls(int indent) override
    {
        cout << string(indent, ' ') << "Folder: " << name << endl;
        for (auto f : fs)
        {
            f->ls(indent + 4);
        }
    }

    void openAll(int indent) override
    {
        cout << string(indent, ' ') << "Opening folder: " << name << endl;
        for (auto f : fs)
        {
            f->openAll(indent + 4);
        }
    }

    int getSize() override
    {
        int size = 0;
        for (auto f : fs)
        {
            size += f->getSize();
        }
        return size;
    }

    string getName() override
    {
        return name;
    }

    bool isFolder() override
    {
        return true;
    }

    FileSystem *changeFolder(const string &name) override
    {
        for (auto f : fs)
        {
            if (f->isFolder() && f->getName() == name)
            {
                return f;
            }
        }
        return nullptr;
    }

    ~Folder()
    {
        for (auto f : fs)
            delete f;
    }
};

int main()
{
    Folder *root = new Folder("open-contribution");

    root->addItem(new File("javascript.js", 1));
    root->addItem(new File("typescript.js", 1));

    Folder *project = new Folder("project");
    project->addItem(new File("index.java", 1));
    project->addItem(new File("main.java", 1));

    root->addItem(project);

    cout << "Total size: " << root->getSize() << endl;
    root->ls(0);
    root->openAll(0);

    FileSystem *cmd = root->changeFolder("project");
    if (cmd)
    {
        cout << "\nChanged directory to: " << cmd->getName() << endl;
        cmd->ls();
    }

    delete root;
    return 0;
}

File: design-pattern/composite-pattern/composite.ts
abstract class FileSystem {
    abstract getName(): string;
    abstract getSize(): number;
    abstract openAll(indent: number): void;
    abstract ls(indent: number): void;
    abstract changeFolder(file: string): FileSystem | null;
    abstract isFolder(): boolean;
};

class File extends FileSystem {
    private name: string;
    private size: number;

    constructor(name: string, size: number) {
        super();
        this.name = name;
        this.size = size;
    };

    getName(): string {
        return this.name;
    };

    getSize(): number {
        return this.size;
    };

    ls(indent: number = 1): void {
        console.log(`${" ".repeat(indent)}File: ${this.name} (${this.size})`);
    };

    openAll(indent: number = 1): void {
        console.log(`${" ".repeat(indent)}Opening file: ${this.name}`);
    };

    changeFolder(f: string): FileSystem | null {
        return null;
    };

    isFolder(): boolean {
        return false;
    };
};

class Folder extends FileSystem {
    private name: string;
    private items: FileSystem[] = [];

    constructor(name: string) {
        super();
        this.name = name;
    };

    addItem(fileSystem: FileSystem): void {
        this.items.push(fileSystem);
    };

    getName(): string {
        return this.name;
    };

    getSize(): number {
        let size: number = 0;

        for (let item of this.items) {
            size += item.getSize();
        };
        return size;
    };

    ls(indent: number = 1): void {
        console.log(`${" ".repeat(indent)}Folder: ${this.name}`);
        for (let item of this.items) {
            item.ls(indent + 2);
        };
    };

    openAll(indent: number = 1): void {
        console.log(`${" ".repeat(indent)}Opening folder: ${this.name}`);
        for (let item of this.items) {
            item.openAll(indent + 2);
        };
    };

    changeFolder(f: string): FileSystem | null {
        for (let item of this.items) {
            if (item.isFolder() && item.getName() === f) {
                return item;
            }
        };
        return null;
    };

    isFolder(): boolean {
        return true;
    };
};


// usage (your style)

let file1 = new Folder("dsa");

file1.addItem(new File("binarySearch.cpp", 1));
file1.addItem(new File("trie.cpp", 1));
file1.addItem(new File("binaryTree.cpp", 1));

let file2 = new Folder("practice");

file2.addItem(new File("sql_practice", 1));
file2.addItem(new File("mongodb_practice", 1));

file1.addItem(file2);

// test
file1.ls(0);
console.log("Total size:", file1.getSize());

let cmd = file1.changeFolder("practice");
cmd?.ls(2);
